input_pipeline/data_extraction.py

```diff
@@ -0,0 +1,82 @@
+from os import makedirs, path
+from typing import Tuple, List, Dict
+from oracle import generate_sequence_of_actions
+import pickle as js
+
+class DataSet(object):
+
+  def __init__(self, lines: List[str]):
+    self.sentences: List[Sentence] = []
+    self.tokens_vocab: Dict[str, int] = {}
+    self.relations_vocab: Dict[str, int] = {}
+
+    entry_lines = []
+    for line in lines:
+      line = line.strip()
+      if line:
+        entry_lines.append(line)
+      elif entry_lines:
+        self.sentences.append(Sentence(entry_lines, self))
+        entry_lines.clear()
+    if entry_lines:
+      self.sentences.append(Sentence(entry_lines, self))
+
+class Sentence(object):
+
+  def __init__(self, lines: List[str], dataset: DataSet):
+    self.tokens: List[int] = [-1]
+    self.heads: List[int] = [-1]
+    self.labels: List[int] = [None]
+
+    for line in lines:
+      line_fields = line.split('\t')
+      if len(line_fields) > 7:
+        token, head, relation = str(line_fields[2]), int(line_fields[6]), str(line_fields[7])
+        if token not in dataset.tokens_vocab.keys():
+          dataset.tokens_vocab.update({token: len(dataset.tokens_vocab.keys())})
+        if relation not in dataset.relations_vocab.keys():
+          dataset.relations_vocab.update({relation: len(dataset.relations_vocab.keys())})
+        self.tokens.append(dataset.tokens_vocab[token])
+        self.heads.append(head)
+        self.labels.append(dataset.relations_vocab[relation])
+
+def load_from_file(file_path: str, cache=True) -> DataSet:
+  """
+  Loads data from file or cache.
+  Args:
+    file_path - the path to the data
+    cache - boolean indicating caching
+  """
+  cached_file_path = file_path + ".cached"
+
+  if cache and path.exists(cached_file_path):
+    print("Using cached data from: " + cached_file_path)
+    with open(cached_file_path, "rb") as cached_file:
+      data = js.load(cached_file)
+  else:
+    print("Extracting data from: " + file_path)
+    with open(file_path, encoding='utf-8') as file:
+      lines = file.readlines()
+      data = DataSet(lines)
+
+      print(cached_file_path)
+      print(path.exists(cached_file_path))
+      if cache and not path.exists(cached_file_path):
+        print("Caching to file: " + cached_file_path)
+        with open(cached_file_path, "wb") as cached_file:
+          js.dump(data, cached_file)
+
+  return data
+
+
+if __name__ == '__main__':
+  file_path = "/home/paul/PycharmProjects/transition-based/mock_data/ro_rrt-ud-test.conllu"
```

Comment:
This path is specific to your workspace (see 'home/paul'). You can use such code snippets for local development but it would be best that the code does not reach the remote repository.
Alternatives:
1. Use local path & document somehwere instructions about downloading the data & placing it where needed
---> this might be better for the training, but would not make sense in the data extraction script necessarily
2. Add (unit) tests instead of running code with 'if __name__ == '__main__'
Since this code is for parsing input files:
---> could have some short input files with simple examples pushed in the repo (that will be used in more tests)
---> could create temporary files during the test run
Not all functions need files though, unit tests could at least be added for those that don't need to open files.